struqture/src/spins/plus_minus_operator.rs

```diff
@@ -0,0 +1,741 @@
+use crate::spins::{PlusMinusProduct, SpinHamiltonian};
+use crate::{
+    OperateOnDensityMatrix, OperateOnState, StruqtureError, StruqtureVersionSerializable,
+    SymmetricIndex, MINIMUM_STRUQTURE_VERSION,
+};
+use num_complex::Complex64;
+use qoqo_calculator::{CalculatorComplex, CalculatorFloat};
+use serde::{Deserialize, Serialize};
+use std::collections::hash_map::{Entry, Iter, Keys, Values};
+use std::collections::HashMap;
+use std::fmt::{self, Write};
+use std::iter::{FromIterator, IntoIterator};
+use std::ops;
+
+/// PlusMinusOperators are combinations of PlusMinusProducts with specific CalculatorComplex coefficients.
+///
+/// This is a representation of sums of pauli products with weightings, in order to build a full hamiltonian.
+///
+/// # Example
+///
+/// ```
+/// use struqture::prelude::*;
+/// use qoqo_calculator::CalculatorComplex;
+/// use struqture::spins::{OperateOnSpins, PlusMinusProduct, PlusMinusOperator};
+///
+/// let mut so = PlusMinusOperator::new();
+///
+/// // Representing the hamiltonian $ 1/2 \sigma_0^{+} \sigma_1^{+} + 1/5 \sigma_0^{z} $
+/// let pp_0x1x = PlusMinusProduct::new().plus(0).plus(1);
+/// let pp_0z = PlusMinusProduct::new().z(0);
+/// so.add_operator_product(pp_0x1x.clone(), CalculatorComplex::from(0.5)).unwrap();
+/// so.add_operator_product(pp_0z.clone(), CalculatorComplex::from(0.2)).unwrap();
+///
+/// // Access what you set:
+/// assert_eq!(so.get(&pp_0x1x), &CalculatorComplex::from(0.5));
+/// assert_eq!(so.get(&pp_0z), &CalculatorComplex::from(0.2));
+/// ```
+///
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(from = "PlusMinusOperatorSerialize")]
+#[serde(into = "PlusMinusOperatorSerialize")]
+pub struct PlusMinusOperator {
+    /// The internal HashMap of PlusMinusProducts and coefficients (CalculatorComplex)
+    internal_map: HashMap<PlusMinusProduct, CalculatorComplex>,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+struct PlusMinusOperatorSerialize {
+    items: Vec<(PlusMinusProduct, CalculatorFloat, CalculatorFloat)>,
+    _struqture_version: StruqtureVersionSerializable,
+}
+
+impl From<PlusMinusOperatorSerialize> for PlusMinusOperator {
+    fn from(value: PlusMinusOperatorSerialize) -> Self {
+        let new_noise_op: PlusMinusOperator = value
+            .items
+            .into_iter()
+            .map(|(key, real, imag)| (key, CalculatorComplex { re: real, im: imag }))
+            .collect();
+        new_noise_op
+    }
+}
+
+impl From<PlusMinusOperator> for PlusMinusOperatorSerialize {
+    fn from(value: PlusMinusOperator) -> Self {
+        let new_noise_op: Vec<(PlusMinusProduct, CalculatorFloat, CalculatorFloat)> = value
+            .into_iter()
+            .map(|(key, val)| (key, val.re, val.im))
+            .collect();
+        let current_version = StruqtureVersionSerializable {
+            major_version: MINIMUM_STRUQTURE_VERSION.0,
+            minor_version: MINIMUM_STRUQTURE_VERSION.1,
```

Comment:
This needs to me minor_version: 1 instead of MINIMUM_STRUQTURE_VERSION.1 because we only add it in 1.1.0